modules/stochastic_tools/src/utils/PolynomialQuadrature.C

```diff
@@ -0,0 +1,269 @@
+
+  Real val = 0.0;
+  for (unsigned int q = 0; q < xq.size(); ++q)
+  {
+    Real prod = wq[q];
+    for (unsigned int i = 0; i < nprod; ++i)
+      prod *= compute(order[i], xq[q], false);
+    val += prod;
+  }
+
+  return val / std::accumulate(wq.begin(), wq.end(), 0.0);
+}
+
+Legendre::Legendre(const UniformDistribution * dist)
+  : Polynomial(), _lower_bound(dist->getLowerBound()), _upper_bound(dist->getUpperBound())
+{
+}
+
+Real
+Legendre::compute(const unsigned int order, const Real x, const bool normalize) const
+{
+  if ((x < _lower_bound) || (x > _upper_bound))
+    ::mooseError("The requested polynomial point is outside of distribution bounds.");
+
+  Real val = legendre(order, x, _lower_bound, _upper_bound);
+  if (normalize)
+    val /= innerProduct(order);
+
+  return val;
+}
+
+void
+Legendre::quadrature(const unsigned int order,
+                     std::vector<Real> & points,
+                     std::vector<Real> & weights) const
+{
+  gauss_legendre(order, points, weights, _lower_bound, _upper_bound);
+}
+
+Real
+legendre(const unsigned int order, const Real x, const Real lower_bound, const Real upper_bound)
+{
+  Real xref = 2 / (upper_bound - lower_bound) * (x - (upper_bound + lower_bound) / 2);
+#ifdef LIBMESH_HAVE_EXTERNAL_BOOST
+  // Using Legendre polynomials from boost library (if available)
+  // https://www.boost.org/doc/libs/1_46_1/libs/math/doc/sf_and_dist/html/math_toolkit/special/sf_poly/legendre.html
+  return boost::math::legendre_p(order, xref);
+#else
+  // Using explicit expression of polynomial coefficients:
+  // P_n(x) = 1/2^n\sum_{k=1}^{floor(n/2)} (-1)^k * nchoosek(n, k) * nchoosek(2n-2k, n) * x^(n-2k)
+  // https://en.wikipedia.org/wiki/Legendre_polynomials
+  Real val = 0;
+  for (unsigned int k = 0; k <= (order % 2 == 0 ? order / 2 : (order - 1) / 2); ++k)
+  {
+    Real coeff =
+        Real(Utility::binomial(order, k)) * Real(Utility::binomial(2 * order - 2 * k, order));
+    Real sgn = (k % 2 == 0 ? 1.0 : -1.0);
+    unsigned int ord = order - 2 * k;
+    val += sgn * coeff * pow(xref, ord);
+  }
+  return val / pow(2.0, order);
+#endif
+}
+
+Hermite::Hermite(const NormalDistribution * dist)
+  : Polynomial(), _mu(dist->getMean()), _sig(dist->getStdDev())
+{
+}
+
+Hermite::Hermite(const BoostNormalDistribution * dist)
+  : Polynomial(), _mu(dist->getMean()), _sig(dist->getStdDev())
+{
+}
+
+Real
+Hermite::compute(const unsigned int order, const Real x, const bool normalize) const
+{
+  Real val = hermite(order, x, _mu, _sig);
+  if (normalize)
+    val /= innerProduct(order);
+  return val;
+}
+
+void
+Hermite::quadrature(const unsigned int order,
+                    std::vector<Real> & points,
+                    std::vector<Real> & weights) const
+{
+  gauss_hermite(order, points, weights, _mu, _sig);
+}
+
+Real
+hermite(const unsigned int order, const Real x, const Real mu, const Real sig)
+{
+  Real xref = (x - mu) / sig;
+#ifdef LIBMESH_HAVE_EXTERNAL_BOOST
```

Comment:
Do we need this if you have the non-boost coded?

Reply:
The boost version is about twice as fast.

Reply:
OK.